core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java

```diff
@@ -0,0 +1,335 @@
+
+    /**
+     * The buffer size used to buffer the content received from the downstream subscriber
+     */
+    private final int maximumBufferSize;
+
+    /**
+     * This publisher is used to send the bytes received from the downstream subscriber's transformers to a
+     * {@link DelegatingBufferingSubscriber} that will buffer a number of bytes specified by the {@code maximumBufferSize}.
+     */
+    private final SimplePublisher<ByteBuffer> publisherToUpstream = new SimplePublisher<>();
+
+    /**
+     * The downstream subscriber that is subscribed to this publisher.
+     */
+    private Subscriber<? super AsyncResponseTransformer<ResponseT, ResponseT>> downstreamSubscriber;
+
+    /**
+     * The amount requested by the downstream subscriber that is still left to fulfill. Updated. when the
+     * {@link Subscription#request(long) request} method is called on the downstream subscriber's subscription. Corresponds to the
+     * number of {@code AsyncResponseTransformer} that will be published to the downstream subscriber.
+     */
+    private final AtomicLong outstandingDemand = new AtomicLong(0);
+
+    /**
+     * tracks if the publisher is currently sending AsyncResponseTransformer to the downstreamSubscriber
+     */
+    private final AtomicBoolean emitting = new AtomicBoolean(false);
+
+    // 'boundedness' info
+    private final long maxElements;
+    private final boolean isBounded;
+    private final AtomicInteger totalEmitted = new AtomicInteger(0);
+    private final AtomicBoolean terminated = new AtomicBoolean(false);
+
+    /**
+     * Creates a publisher bounded by the {@code maxElements} provided. As per reactive stream specification, this publisher will
+     * be considered 'unbounded' if {@code maxElements == LONG.MAX_VALUE}.
+     */
+    public SplittingTransformer(AsyncResponseTransformer<ResponseT, ResultT> upstreamResponseTransformer,
+                                long bufferSize, CompletableFuture<ResultT> returnFuture,
+                                long maxElements) {
+        this.upstreamResponseTransformer = Validate.paramNotNull(upstreamResponseTransformer, "asyncRequestBody");
+        this.returnFuture = Validate.paramNotNull(returnFuture, "returnFuture");
+        this.maximumBufferSize = NumericUtils.saturatedCast(bufferSize);
+        this.maxElements = maxElements;
+        this.isBounded = maxElements != Long.MAX_VALUE;
+    }
+
+    /**
+     * Creates an unbounded publisher.
+     */
+    public SplittingTransformer(AsyncResponseTransformer<ResponseT, ResultT> upstreamResponseTransformer,
+                                long bufferSize, CompletableFuture<ResultT> returnFuture) {
+        this.upstreamResponseTransformer = Validate.paramNotNull(upstreamResponseTransformer, "asyncRequestBody");
+        this.returnFuture = Validate.paramNotNull(returnFuture, "returnFuture");
+        this.maximumBufferSize = NumericUtils.saturatedCast(bufferSize);
```

Comment:
Can we support long?

Reply:
`SplitAsyncResponseTransformer` is just a holder for the publisher and a completable future, the class implementing `AsyncResponseTransformer` is  `SplittingTransformer.IndividualTransformer`

Reply:
we might need to allocate a ByteBuffer of that size in the case we receive a bigger `ByteBuffer` from the publisher of the individual request body. See [DelegatingBufferingSubscriber#L89](https://github.com/aws/aws-sdk-java-v2/pull/4826/files#diff-d03cd0cb555b1f4448dd14c8559f557fd9ae1b0a0e74b6e4b640441f7cf93058R89)